app/scripts/lib/multichain-api/MultichainMiddlewareManager.ts

```diff
@@ -0,0 +1,87 @@
+import { JsonRpcMiddleware } from 'json-rpc-engine';
+import { Scope } from './scope';
+
+// Extend JsonRpcMiddleware to include the destroy method
+export type ExtendedJsonRpcMiddleware = JsonRpcMiddleware<unknown, unknown> & {
+  destroy?: () => void;
+};
+
+type MiddlewareByScope = Record<Scope, ExtendedJsonRpcMiddleware>;
+
+type MiddlewareManager = {
+  removeMiddleware: (scope: Scope, domain: string) => void;
+  removeAllMiddleware: () => void;
+  addMiddleware: (
+    scope: Scope,
+    domain: string,
+    middleware: ExtendedJsonRpcMiddleware,
+  ) => void;
+  middleware: ExtendedJsonRpcMiddleware;
+};
+
+export default function createMultichainMiddlewareManager(): MiddlewareManager {
```

Comment:
should this be rewritten as a class?

Reply:
it was a class but I couldn't figure out a good way to add `.destroy` to the middleware function

Reply:
```
import { JsonRpcMiddleware } from 'json-rpc-engine';
import { Scope } from './scope';

// Extend JsonRpcMiddleware to include the destroy method
export type ExtendedJsonRpcMiddleware = JsonRpcMiddleware<unknown, unknown> & {
  destroy?: () => void;
};

type MiddlewareByScope = Record<Scope, ExtendedJsonRpcMiddleware>;

export default class MultichainMiddlewareManager {
  private middlewareCountByDomainAndScope: {
    [scope: string]: { [domain: string]: number };
  } = {};

  private middlewaresByScope: MiddlewareByScope = {};

  public removeAllMiddleware(): void {
    for (const [scope, domainObject] of Object.entries(
      this.middlewareCountByDomainAndScope,
    )) {
      for (const domain of Object.keys(domainObject)) {
        this.removeMiddleware(scope as Scope, domain);
      }
    }
  }

  public addMiddleware(
    scope: Scope,
    domain: string,
    middleware: ExtendedJsonRpcMiddleware,
  ): void {
    this.middlewareCountByDomainAndScope[scope] =
      this.middlewareCountByDomainAndScope[scope] || {};
    this.middlewareCountByDomainAndScope[scope][domain] =
      this.middlewareCountByDomainAndScope[scope][domain] || 0;
    this.middlewareCountByDomainAndScope[scope][domain] += 1;
    if (!this.middlewaresByScope[scope]) {
      this.middlewaresByScope[scope] = middleware;
    }
  }

  public removeMiddleware(scope: Scope, domain: string): void {
    if (this.middlewareCountByDomainAndScope[scope]?.[domain]) {
      this.middlewareCountByDomainAndScope[scope][domain] -= 1;
      if (this.middlewareCountByDomainAndScope[scope][domain] === 0) {
        delete this.middlewareCountByDomainAndScope[scope][domain];
      }
      if (
        Object.keys(this.middlewareCountByDomainAndScope[scope]).length === 0
      ) {
        delete this.middlewareCountByDomainAndScope[scope];
        delete this.middlewaresByScope[scope];
      }
    }
  }

  public middleware: ExtendedJsonRpcMiddleware = (req, res, next, end) => {
    const r = req as unknown as { scope: string };
    const { scope } = r;
    if (typeof this.middlewaresByScope[scope] === 'function') {
      this.middlewaresByScope[scope](req, res, next, end);
    } else {
      next();
    }
  };

  constructor() {
    this.middleware.destroy = this.removeAllMiddleware.bind(this);
  }
}
```

Reply:
this ☝️ should work?